### PR TITLE
Method Ordering Proposal

### LINKED ISSUE #12: pragma marks and file structure
Hi,

I'm pretty sure you're using pragma marks a lot.

I'm curious what's your best practise to structure code then.

Here's an example from us (@nxtbgthng):

We usually structure our implementation files like so like so (let's call this one Class.m):

``` obj-c
/* copy right comment if necessary */

#import <Framework/Framework.h>

#import "OtherHeader.h"
#import "OtherHeaderInSemanticalGroup.h"

#import "NextHeader.h"

#import "Class.h" // we import the implementation files header last


// place for consts and defines if you can't avoid them


@interface Class () <SecretNotVisibleFromHeaderProtocol>
@property (readonly) NSString *text;
@end


@implementation Class

#pragma mark Class Methods

+ (void)load;
{
}

#pragma mark Lifecycle

- (id)init;
{
    ....
}

- (void)dealloc;
{
    ....
}


#pragma mark Accessors

@synthesize text = _text; // readonly properties don't come with an ivar by default
- (NSString *)text
{
    if (!_text) {
        _text = @"text";
    }
    return _text;
}


#pragma mark SecretNotVisibleFromHeaderProtocol

- (void)methodDefinedInTheProtocol;
{
}


#pragma mark UIResponder

- (void)methodOverridenFromUIResponder;
{
}


#pragma mark Private Helper

- (CGRect)rectFromStuff:(id)stuff;
{
}

@end


#pragma mark -
#pragma mark Stuff private to this class

// Put private implementation blocks and C functions here.
// They need to be declared on top of the file though and might make things a little more messy
```

So key points are:
- use pragma marks to group methods
- when subclassing or implementing a protocol group the methods that do so under the name of the super class or protocol
- always leave two empty lines above a pragma mark (we found them to be more visible that way)
- use a `Lifecycle` section to group all init and dealloc
- use a `Class Methods` section to group all static methods
- use a `Accessors` section to group all custom getters and setters
- keep `@synthesize`/`@dynamic` close to the getter/setter

Thanks for sharing your style guide. It's pretty close to how we do it which is a good sign for both of us ;)

-ullrich

### LINKED ISSUE #44: Declarations/definitions of class method should precede instance methods
I find it oddly disconcerting if class methods are interspersed with instance methods in interfaces and implementations

### PR DESCRIPTION
**Pull request by mattbischoff**
Here’s a proposal on how to handle the Method Ordering conundrum that’s come up in #56 and #44. Would love to hear what everyone thinks here.

I do think it makes sense to have a consistent way to do this, and this way has worked for me, but totally open to other ideas.

Closes #44.
Closes #12.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -352,6 +351,51 @@ Private properties should be declared in class extensions (anonymous categories)
 @end
 ```
 
+## Method Ordering
+
+Method implementations are grouped and ordered by class inheritance, then by protocol conformance with system protocols proceeding custom protocols. A `pragma mark` is used to separate each group.
```
Comment (cdzombak) [inline README.md]: > with system protocols proceeding custom protocols

I am not sure I agree with specifying this explicitly. Often my custom protocols feel, for lack of a better term, more "important" or more "domain-specific" than the UIKit et al. protocols, which tend to be more generic and feel more like boilerplate (obviously generalizing here).

Aside from this, I'm on board with this PR; @mattbischoff, what would you think about eliminating this one phrase? I don't think that change diminishes the value of the guideline.

**Comment by cdzombak** [comment]
_reposting since I accidentally added this comment on #12 instead of here:_

> Closes #44.

Unless I am missing something, this PR does not address ordering of class vs. instance methods.

**Comment by mattbischoff** [comment]
Looks like I owe this some love.

**Comment by mattbischoff** [comment]
More discussion going on at #120. Will fork off that and make changes if necessary.

**Pull request closed without merging**